provider/exoscale.go

```diff
@@ -0,0 +1,201 @@
+package provider
+
+import (
+	"strings"
+
+	"github.com/exoscale/egoscale"
+	"github.com/kubernetes-incubator/external-dns/endpoint"
+	"github.com/kubernetes-incubator/external-dns/plan"
+	log "github.com/sirupsen/logrus"
+)
+
+// EgoscaleClientI for replaceable implementation
+type EgoscaleClientI interface {
+	GetRecords(string) ([]egoscale.DNSRecord, error)
+	GetDomains() ([]egoscale.DNSDomain, error)
+	CreateRecord(string, egoscale.DNSRecord) (*egoscale.DNSRecord, error)
+	DeleteRecord(string, int64) error
+}
+
+// ExoscaleProvider initialized as dns provider with no records
+type ExoscaleProvider struct {
+	domain         DomainFilter
+	client         EgoscaleClientI
+	filter         *zoneFilter
+	OnApplyChanges func(changes *plan.Changes)
+}
+
+// ExoscaleOption for Provider options
+type ExoscaleOption func(*ExoscaleProvider)
+
+// NewExoscaleProvider returns ExoscaleProvider DNS provider interface implementation
+func NewExoscaleProvider(endpoint, apiKey, apiSecret string, opts ...ExoscaleOption) *ExoscaleProvider {
+	client := egoscale.NewClient(endpoint, apiKey, apiSecret)
+	return NewExoscaleProviderWithClient(endpoint, apiKey, apiSecret, client, opts...)
+}
+
+// NewExoscaleProviderWithClient returns ExoscaleProvider DNS provider interface implementation (Client provided)
+func NewExoscaleProviderWithClient(endpoint, apiKey, apiSecret string, client EgoscaleClientI, opts ...ExoscaleOption) *ExoscaleProvider {
+	ep := &ExoscaleProvider{
+		filter:         &zoneFilter{},
+		OnApplyChanges: func(changes *plan.Changes) {},
+		domain:         NewDomainFilter([]string{""}),
+		client:         client,
+	}
+	for _, opt := range opts {
+		opt(ep)
+	}
+	return ep
+}
+
+func (ep *ExoscaleProvider) getZones() (map[int64]string, error) {
+	dom, err := ep.client.GetDomains()
+	if err != nil {
+		return nil, err
+	}
+
+	zones := map[int64]string{}
+	for _, d := range dom {
+		zones[d.ID] = d.Name
+	}
+	return zones, nil
+}
+
+// ApplyChanges simply modifies DNS via exoscale API
+func (ep *ExoscaleProvider) ApplyChanges(changes *plan.Changes) error {
+	ep.OnApplyChanges(changes)
+
+	zones, err := ep.getZones()
+	if err != nil {
+		return err
+	}
+
+	for _, epoint := range changes.Create {
+		if ep.domain.Match(epoint.DNSName) {
+			if zoneID, name := ep.filter.EndpointZoneID(epoint, zones); zoneID != 0 {
+				rec := egoscale.DNSRecord{
+					Name:       name,
+					RecordType: epoint.RecordType,
+					TTL:        int(epoint.RecordTTL),
+					Content:    epoint.Targets[0],
+				}
+				_, err := ep.client.CreateRecord(zones[zoneID], rec)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	for _, epoint := range changes.UpdateNew {
```

Comment:
I must admit that for my use case it seems to be sufficient to only support create and delete. Is it mandatory?

Reply:
I'm just only asking because I never saw that someone wouldn't want to update their records. Could there people which might need it?

If you don't use it at all I'm not seeing any point looping over updates which doesn't do anything execept logging.